includes/data-stores/class-wc-admin-reports-customers-data-store.php

```diff
@@ -0,0 +1,399 @@
+			if ( ! empty( $query_args[ $exact_match_param ] ) ) {
+				$where_clauses[] = $wpdb->prepare(
+					"{$order_product_lookup_table}.{$exact_match_param} = %s",
+					$query_args[ $exact_match_param ]
+				); // WPCS: unprepared SQL ok.
+			}
+		}
+
+		$numeric_params = array(
+			'orders_count'    => '%d',
+			'total_spend'     => '%f',
+			'avg_order_value' => '%f',
+		);
+
+		foreach ( $numeric_params as $numeric_param => $sql_format ) {
+			$subclauses = array();
+			$min_param  = $numeric_param . '_min';
+			$max_param  = $numeric_param . '_max';
+
+			if ( isset( $query_args[ $min_param ] ) ) {
+				$subclauses[] = $wpdb->prepare(
+					"{$order_product_lookup_table}.{$numeric_param} >= {$sql_format}",
+					$query_args[ $min_param ]
+				); // WPCS: unprepared SQL ok.
+			}
+
+			if ( isset( $query_args[ $max_param ] ) ) {
+				$subclauses[] = $wpdb->prepare(
+					"{$order_product_lookup_table}.{$numeric_param} <= {$sql_format}",
+					$query_args[ $max_param ]
+				); // WPCS: unprepared SQL ok.
+			}
+
+			if ( $subclauses ) {
+				$where_clauses[] = '(' . implode( ' AND ', $subclauses ) . ')';
+			}
+		}
+
+		if ( $where_clauses ) {
+			$preceding_match = empty( $sql_query_params['where_time_clause'] ) ? ' AND ' : " {$match_operator} ";
+			$sql_query_params['where_clause'] = $preceding_match . implode( " {$match_operator} ", $where_clauses );
+		}
+
+		return $sql_query_params;
+	}
+
+	/**
+	 * Returns the report data based on parameters supplied by the user.
+	 *
+	 * @param array $query_args  Query parameters.
+	 * @return stdClass|WP_Error Data.
+	 */
+	public function get_data( $query_args ) {
+		global $wpdb;
+
+		$table_name = $wpdb->prefix . self::TABLE_NAME;
+
+		// These defaults are only partially applied when used via REST API, as that has its own defaults.
+		$defaults = array(
+			'per_page' => get_option( 'posts_per_page' ),
+			'page'     => 1,
+			'order'    => 'DESC',
+			'orderby'  => 'date_registered',
+			'fields'   => '*',
+		);
+		$query_args = wp_parse_args( $query_args, $defaults );
+
+		$cache_key = $this->get_cache_key( $query_args );
+		$data      = wp_cache_get( $cache_key, $this->cache_group );
+
+		if ( false === $data ) {
+			$data = (object) array(
+				'data'    => array(),
+				'total'   => 0,
+				'pages'   => 0,
+				'page_no' => 0,
+			);
+
+			$selections       = $this->selected_columns( $query_args );
+			$sql_query_params = $this->get_sql_query_params( $query_args );
+
+			$db_records_count = (int) $wpdb->get_var(
+				"SELECT COUNT(*)
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+				"
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			$total_pages = (int) ceil( $db_records_count / $sql_query_params['per_page'] );
+			if ( $query_args['page'] < 1 || $query_args['page'] > $total_pages ) {
+				return $data;
```

Comment:
When we return early, do we still want to set the cache like we do in line 307?

Reply:
I don't think so - in this case, a page that is outside the possible result set is being requested (page `0` or a page beyond the total number of pages). We haven't actually queried for data yet, just a count, so there's really nothing to cache.